packages/flutter/lib/src/rendering/object.dart

```diff
@@ -4583,133 +4328,609 @@ mixin RelayoutWhenSystemFontsChangeMixin on RenderObject {
-    required this.siblingMergeGroups,
-  });
-
-  final List<List<_InterestingSemanticsFragment>> siblingMergeGroups;
+class _IncompleteSemanticsFragmentProvider extends _InterestingSemanticsFragmentProvider {
+  _IncompleteSemanticsFragmentProvider(this.fragment);
 
-  @override
-  void addAll(Iterable<_InterestingSemanticsFragment> fragments) {
-    mergeUpFragments.addAll(fragments);
-  }
+  final _IncompleteSemanticsFragment fragment;
 
   @override
-  final List<_InterestingSemanticsFragment> mergeUpFragments = <_InterestingSemanticsFragment>[];
+  _InterestingSemanticsFragment getFragment() => fragment;
 }
 
-/// A [_SemanticsFragment] that describes which concrete semantic information
-/// a [RenderObject] wants to add to the [SemanticsNode] of its parent.
-///
-/// Specifically, it describes which children (as returned by [compileChildren])
-/// should be added to the parent's [SemanticsNode] and which [config] should be
-/// merged into the parent's [SemanticsNode].
-abstract class _InterestingSemanticsFragment extends _SemanticsFragment {
-  _InterestingSemanticsFragment({
-    required RenderObject owner,
-    required super.dropsSemanticsOfPreviousSiblings,
-  }) : _ancestorChain = <RenderObject>[owner];
-
-  /// The [RenderObject] that owns this fragment (and any new [SemanticsNode]
-  /// introduced by it).
-  RenderObject get owner => _ancestorChain.first;
+/// A wrapper class for a [RenderObject] that provides semantics related
+/// properties and compilations.
+class _RenderObjectSemantics extends _InterestingSemanticsFragmentProvider {
+  _RenderObjectSemantics(this.renderObject);
 
-  final List<RenderObject> _ancestorChain;
+  /// The owning rendering object for this object.
+  final RenderObject renderObject;
 
-  /// The children to be added to the parent.
-  ///
-  /// See also:
-  ///
-  ///  * [SemanticsNode.parentSemanticsClipRect] for the source and definition
-  ///    of the `parentSemanticsClipRect` argument.
-  ///  * [SemanticsNode.parentPaintClipRect] for the source and definition
-  ///    of the `parentPaintClipRect` argument.
-  ///  * [SemanticsNode.elevationAdjustment] for the source and definition
-  ///    of the `elevationAdjustment` argument.
-  void compileChildren({
-    required Rect? parentSemanticsClipRect,
-    required Rect? parentPaintClipRect,
-    required double elevationAdjustment,
-    required List<SemanticsNode> result,
-    required List<SemanticsNode> siblingNodes,
-  });
+  // We use an additional boolean to keep track of cache dirtiness since we can't
+  // remove `_cachedSemanticFragment` outright when RenderObject is marked
+  // dirty. It is possible a parent rendering object attempt to access a child
+  // rendering object's _semanticsFragmentProvider while it is still dirty. In
+  // this case, we need the previous _cachedSemanticFragment's information ,e.g.
+  // its ancestor chain, to compute new semantics fragment.
+  bool isDirty = true;
```

Comment:
What exactly is dirt when this flag is set? What contains outdated information when this is true? e.g. is it OK to rely on the information on cachedSemanticsNode or _cachedSemanticsConfiguration? What needs to be called to clear this flag?

Reply:
Can you add the answers to these questions to the comment?